### PR TITLE
podman image scp remote support & podman image scp tagging

### PR DESCRIPTION
**Pull request by cdoern**
add support for podman-remote image scp as well as direct access via the API. This entailed
a full rework of the layering of image scp functions as well as the usual API plugging and type creation

also, implemented podman image scp tagging. which makes the syntax much more readable and allows users t tag the new image
they are loading to the local/remote machine:

allow users to pass a "new name" for the image they are transferring
`podman tag` as implemented creates a new image im `image list` when tagging, so this does the same
meaning that when transferring images with tags, podman on the remote machine/user will load two images
ex: `podman image scp computer1::alpine computer2::foobar` creates alpine:latest and localhost/foobar on the remote host

implementing tags means removal of the buggy syntax. In the currently released podman image scp, the user can either specify
`podman image scp source::img dest::` or `podman image scp dest:: source::img`. However, with tags this task becomes really hard to check
which is the image (src) and which is the new tag (dst). Removal of that streamlines the arg parsing process

Signed-off-by: cdoern <cbdoer23@g.holycross.edu>

#### Does this PR introduce a user-facing change?

```release-note
podman image scp API support added as well as image tagging. This fixes a bug in the syntax allowing users to specify a destination before a source. The API endpoint takes a source, destination, and quiet boolean all in the query itself.

```

### CODE REVIEW

File: pkg/domain/infra/abi/images.go
Comment (rhatdan) [inline pkg/domain/infra/abi/images.go:905]: Nit just 
return ioutil.WriteFile(...)

**Comment by edsantiago** [comment]
@cdoern could I encourage you to rebase onto main any time you do a new push?
```console
$ git checkout main
$ git pull -r (or however you update main)
$ git checkout my-branch
$ git rebase main
! NOW you can git-push-force!
```
Advantages:
* You pick up the latest flake fixes, which means
  * you and everyone waste less time pressing Re-run
  * /me wastes less time looking at flake logs, wondering why this-or-that flake is recurring after being fixed a week ago
* and, there is less likelihood of a future collision. (Not all collisions cause merge conflicts. Some collisions happen because of changes in functionality, not blocks of code. These cannot be detected automatically). The closer your PR is to current main, the less chance there is of that happening.

Thank you!

**Comment by cdoern** [comment]
sure @edsantiago sorry for the inconvenience. I usually do:
`git checkout <my branch>`
`git fetch upstream main`
`git rebase upstream/main`
`git push -f origin <my branch>`

This has the same effect I think right?

**Comment by edsantiago** [comment]
Hmmm, it should, but if you look at your git tree you'll see that this PR is coming from a72b2402c which is a week old. Yet you pushed it 3 hours ago. Did you not follow your usual process? You might want to look at those step-by-step and see where the breakdown is. Thanks again!

File: pkg/domain/utils/utils.go
Comment (mheon) [inline pkg/domain/utils/utils.go:64]: Can you throw all the SCP functions in a separate file, same package (`pkg/domain/utils/scp.go` maybe?)

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL

**Comment by cdoern** [comment]
@edsantiago any idea why you changed adding `stop_registry` to https://github.com/containers/podman/blob/bce97a3b5dd1477c74232b49d246491aa289a995/test/apiv2/12-imagesMore.at seem to be breaking the API v2 tests here?

File: pkg/api/handlers/libpod/images.go
Comment (jwhonce) [inline pkg/api/handlers/libpod/images.go:693]: @cdoern It appears in the ExecuteTransfer() call tree, that reading from stdin is required. There is no support there for reading from the remote host. The user is queried for the password when no key is configured.

File: pkg/api/server/register_images.go
Comment (jwhonce) [inline pkg/api/server/register_images.go:1630]:
```suggestion
	// swagger:operation POST /libpod/images/scp/{name} libpod ImageScpLibpod
	// ---
	// tags:
	//  - images
	// summary: Copy an image from one host to another
	// description: Copy an image from one host to another
	// parameters:
	//   - in: path
	//     name: name
	//     required: true
	//     description: source connection/image
	//     type: string
```

File: pkg/bindings/images/images.go
Comment (jwhonce) [inline pkg/bindings/images/images.go:350]:
```suggestion
func Scp(ctx context.Context, source, destination string, options ScpOptions) (reports.ScpReport, error) {
```

File: pkg/bindings/images/types.go
Comment (jwhonce) [inline pkg/bindings/images/types.go:192]:
```suggestion
```

File: pkg/bindings/images/images.go
Comment (jwhonce) [inline pkg/bindings/images/images.go:362]: Update code to format url to use source in path vs. query parameter.

**Review by jwhonce — CHANGES_REQUESTED** [review]
@cdoern There is a gotcha here, that code you have moved into the domain is currently trying to query the user via stdin. This will not work in the remote use case.

You could change that code to use a call-back that is configured in cmd/podman/connection to do the query and in the API handler the callback would return an error. WDYT?

**Comment by edsantiago** [comment]
@cdoern followup: I can reproduce this in CI using the "rerun terminal" trick that @cevich demoed yesterday (THANK YOU @cevich!). I cannot understand what is going on. This has been the only thing I've worked on this morning, it is my entire focus (except for two minutes on a different PR review). I am baffled, but I am continuing to focus 100% on it. Will keep you posted.

**Comment by cdoern** [comment]
thanks for the update @edsantiago !!! I @ you in another PR (#14299) where the same thing is happening. Let me know if I can help in any way

File: test/apiv2/12-imagesMore.at
Comment (cdoern) [reply test/apiv2/12-imagesMore.at:63]: @edsantiago what about the same failure on a sep PR? https://github.com/containers/podman/runs/6815997255

File: pkg/api/handlers/libpod/images.go
Comment (cdoern) [reply pkg/api/handlers/libpod/images.go:693]: @jwhonce the only time where stdin is needed is the intra machine user transfer. This use case is not supported on the remote client for a bunch of reasons. Given that, I think i can keep this structure as is.

File: pkg/api/server/register_images.go
Comment (cdoern) [reply pkg/api/server/register_images.go:1630]: I will put this in the path but keep it as source as it can be either an image or a ssh destination (and the native docs uses source/dest lingo)

File: pkg/bindings/images/images.go
Comment (cdoern) [reply pkg/bindings/images/images.go:350]: why this change?

File: pkg/bindings/images/types.go
Comment (cdoern) [reply pkg/bindings/images/types.go:192]: shouldn't just source be deleted here?

File: pkg/api/server/register_images.go
Comment (cdoern) [reply pkg/api/server/register_images.go:1630]: oh do I have to call this name in order to put it in the path/pull it out of the path @jwhonce ?

File: pkg/domain/infra/tunnel/images.go
Comment (cdoern) [reply pkg/domain/infra/tunnel/images.go:381]: not sure if that is possible since the native scp function will not be returning a string since it gets printed via exec.Command

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL

File: pkg/domain/infra/tunnel/images.go
Comment (jwhonce) [inline pkg/domain/infra/tunnel/images.go:367]:
```suggestion
func (ir *ImageEngine) Scp(ctx context.Context, source, destination string, parentFlags []string, quiet bool) error {
```
Rather than hide source and destination in args, call them out by name.

Comment (cdoern) [reply pkg/domain/infra/tunnel/images.go:367]: for the native implementation I need to put them back in an array for parsing if I make this change.

I will change the name of the array here to make it more obvious what we are dealing with is that ok?

**Comment by cdoern** [comment]
@jwhonce @containers/podman-maintainers PTAL. I have other scp issues and commits I need to work on after this merges.

File: pkg/domain/infra/abi/images.go
Comment (jwhonce) [inline pkg/domain/infra/abi/images.go:897]:
```suggestion
		if !errors.Is(err, fs.ErrExist) {
```
https://pkg.go.dev/os#IsExist

File: pkg/domain/infra/tunnel/images.go
Comment (jwhonce) [inline pkg/domain/infra/tunnel/images.go:367]: Using an array for source and destination seems smelly. Why not use parameters for source and destination?

File: pkg/domain/utils/utils.go
Comment (jwhonce) [inline pkg/domain/utils/utils.go:6]: change seems gratuitous

**Review by jwhonce — COMMENTED** [review]
Small items

**Comment by cdoern** [comment]
@jwhonce @containers/podman-maintainers all tests green here, PTAL and merge

**Comment by jwhonce** [comment]
LGTM

**Pull request merged**
